gandy/models/models.py

```diff
@@ -0,0 +1,358 @@
+                array of prediction uncertainties of targets withthe same length 
+                    as Xs
+                (optional) array of flags of uncertain predictions higher than thr-
+                    eshhold of same length as Xs
+        """
+        ## pseudocode
+        #. check X data with check function
+        #. run _predict to return predictions and uncertainties
+        #. if threshhold, return predictions, uncertainties, and flags
+        return predictions, uncertainties, flags
+    
+    def _predict(self, 
+                 Xs: Array, 
+                 *args, 
+                 **kwargs):
+        """Make predictions on a set of data and return predictions and uncertain-
+        ty arrays.
+        
+        Must be implimented by child class. Makes predictions on data using
+        model at self.model and any other stored objects.
+        
+        Args:
+            Xs (ndarray):
+                Example data to make predictions on.
+            *args:
+                Positional arguments to be defined by child.
+            **kwargs: 
+                Keyword arguments for predicting.
+                
+        Returns:
+            tuple of ndarray:
+                array of predictions of targets with the same length as Xs
+                array of prediction uncertainties of targets withthe same length 
+                    as Xs
+        """
+        ## psuedocode
+        #. raise not implimented
+        #. set pred, unc to None
+        return predictions, uncertainties
+    
+    def score(self, 
+              Xs: Iterable,
+              Ys: Iterable,
+              metric: Union[str, Callable], 
+              **kwargs) -> Tuple[float, Array]:
+        """Make predictions and score the results according to a defined metric.
+        
+        For a set of labeled example data, use the the defined `_predict`
+        method to make predictions on the data. Then, compare them to the true
+        labels according to a desired metric.
+        
+        Args:
+            Xs (Iterable): 
+                Examples data to make predictions on.
+            Ys (Iterable): 
+                Labels of data.
+            metric (str): 
+                Metric to use, a key in UncertaintyModel.metrics or a metric object
+                that takes as input true, predicted, and uncertainty values.
+            **kwargs: keyword arguments to pass to `_predict`
+            
+        Returns:
+            float:
+                Total score according to metric.
+            ndarray:
+                Score array for each prediction.
+        """
+        ## pseudocode
+        #. if statement to get metric object from metrics or specified
+        #. else raise undefined metric
+        #. check data
+        #. predictions, uncertainties = execute self.predict on Xs
+        #. pass predictions, uncertainties to metric get back costs
+        return cost, costs
+        
+    def save(self, 
```

Comment:
Can we make this a child defined method instead eg. predict. Raise not implemented in parent